SC-687-crosscontract-invoke

potan opened the pull request.

Commit:
SC-687-crosscontract-invoke

Commit:
SC-687-crosscontract-invoke
Fix some tests

Commit:
SC-687-crosscontract-invoke
Test for nested invoke

Commit:
SC-687-crosscontract-invoke
AttachedPayments in callbacks.

Commit:
SC-687-crosscontract-invoke
Scripts invocations limit.

Commit:
SC-687-crosscontract-invoke
Calculate runsLimit from fee (+1 for save old error messages)

Commit:
SC-687-crosscontract-invoke
Limiting the depth of the call separately from the script launches.

Commit:
SC-687-crosscontract-invoke
Use old error message.

Commit:
Merge remote-tracking branch 'origin' into SC-687-crosscontract-invoke

Commit:
Merge remote-tracking branch 'origin' into SC-687-crosscontract-invoke

Commit:
SC-687-crosscontract-invoke
Merging fix

Commit:
Merge remote-tracking branch 'origin' into SC-687-crosscontract-invoke

Commit:
SC-687-crosscontract-invoke
Representation of return value.

Commit:
SC-687-crosscontract-invoke
fix benchmark

Commit:
SC-687-crosscontract-invoke
Set complexity to 30

Commit:
SC-687-crosscontract-invoke
Issue action

Commit:
SC-687-crosscontract-invoke
Smart asset support

Commit:
Merge branch 'master' into SC-687-crosscontract-invoke

Commit:
Merge remote-tracking branch 'origin' into SC-687-crosscontract-invoke

Commit:
Merge remote-tracking branch 'origin' into SC-687-crosscontract-invoke

Commit:
SC-687-crosscontract-invoke
Fix merging

Commit:
SC-687-crosscontract-invoke
Fix Invoke complexity.

Commit:
SC-687-crosscontract-invoke
Call default function (+test)
Call by alias

Commit:
SC-687-crosscontract-invoke
Invoke by alist (+test)

Commit:
SC-687-crosscontract-invoke
Fix types

Commit:
SC-687-crosscontract-invoke
Test for avaliability of payment.

Commit:
SC-687-crosscontract-invoke
Test invoke fields (caller, callerPublicKey).

Commit:
Merge remote-tracking branch 'origin/master' into SC-687-crosscontract-invoke

Commit:
SC-687-crosscontract-invoke
V5 script support

Commit:
SC-687-crosscontract-invoke
Invoke function compilation test

Commit:
SC-687-crosscontract-invoke
Cleaning fee calculation in tests.

mrkraft committed:
Add feature 16 support.

Commit:
SC-687-crosscontract-invoke
Add invokes placeholder to the ScriptResult.

Commit:
Merge branch 'SC-687-crosscontract-invoke' of ssh://github.com/wavesplatform/Waves into SC-687-crosscontract-invoke

Commit:
SC-687-crosscontract-invoke
collect invokes

xrtm000 committed:
Test balances

xrtm000 committed:
Format InvokeScriptTransactionDiffTest

xrtm000 committed:
Merge branch 'SC-687-crosscontract-invoke' of github.com:wavesplatform/Waves into SC-687-crosscontract-invoke

msmolyakov committed:
SC-687 Tests for Ride V5 activation and interoperability with InvokeScriptTx versions (#3326)

Commit:
SC-687 Self payment tests (#3327)

* self payment dapp to dapp tests

* fix

mrkraft committed:
Sc 687 state update tests (#3328)

* SC-687 State update tests

* SC-687 State update tests

Commit:
SC-687-crosscontract-invoke
Check payments with smart-assets.

Commit:
Merge branch 'SC-687-crosscontract-invoke' of ssh://github.com/wavesplatform/Waves into SC-687-crosscontract-invoke

Commit:
SC-687-crosscontract-invoke
Fix warning

Commit:
SC-687-crosscontract-invoke
support invokes stateChanges

Commit:
SC-687-crosscontract-invoke
stateChanges protobuf support

Commit:
SC-687-crosscontract-invoke
Test for stateChanges.invokes

xrtm000 commented on node/src/test/scala/com/wavesplatform/state/diffs/ci/InvokeScriptTransactionDiffTest.scala:
```diff
@@ -2172,9 +2181,1221 @@ class InvokeScriptTransactionDiffTest
+        invoker <- accountGen
+        service <- accountGen
+        ts      <- timestampGen
+        fee     <- ciFee(1)
```
Why is it enough despite issuing via action?
See line 2330:
https://github.com/wavesplatform/Waves/pull/3316/files#diff-7a87dd14a82b2cd0e58fddefc26230665e583d933ab5ba4da3ea943f13f4a7f1R2330

potan replied:
I use NFT

xrtm000 replied:
If set Issue action as non-NFT there still will be no an error

potan replied:
Fixed

xrtm000 commented on node/src/main/scala/com/wavesplatform/state/diffs/invoke/InvokeScriptTransactionDiff.scala:
```diff
@@ -100,23 +102,19 @@ object InvokeScriptTransactionDiff {
+                  pk,
+                  dAppAddress,
+                  runsLimit,
+                  13
```
What is it? Should be it tested?

potan replied:
It limits nested Invoke. Tests "Infilite recurcive crosscontract call".

xrtm000 commented on node-it/src/test/scala/com/wavesplatform/it/api/model.scala:
```diff
@@ -416,14 +416,32 @@ object AssetPairResponse {
+}
+
+case class Invokation(dApp: String, call: Call, payments: Seq[AttachedPayment], stateChanges: StateChangesDetails)
+object Invokation {
```
```suggestion
object Invocation {
```

xrtm000 commented on node/src/main/scala/com/wavesplatform/state/InvokeScriptResult.scala:
```diff
@@ -56,14 +67,23 @@ object InvokeScriptResult {
-    override def combine(x: InvokeScriptResult, y: InvokeScriptResult): InvokeScriptResult =
-      InvokeScriptResult(x.data ++ y.data, x.transfers ++ y.transfers)
+    override def combine(x: InvokeScriptResult, y: InvokeScriptResult): InvokeScriptResult = {
+      InvokeScriptResult(/*x.data ++ y.data, x.transfers ++ y.transfers,*/ invokes = x.invokes ++ y.invokes)
```
Should it be saved unchanged?

xrtm000 committed:
SC Mix continuation with sync dApps (#3331)

* Optimize imports

* Correct PseudoTransactionSuite

* Correct integration tests

* Support /transactions/address/

* Better naming

* Support fail of Continuation Transaction

* Correct fail of Continuation Transaction

* Test fail of Continuation Transaction

* Format code

* Comment unimplemented code

* Fix extension method call

* Add fee parameters for Continuation transaction

* Fill fee for Continuation transaction

* Check fee in test

* Support fill of fee for failed continuations

* Support /address/transactions for failed continuations

* Correct and refactor construct of failed continuation diff

* Better error on absence of binding transaction

* Correct extracting data functions docs for V5

* Correct extraFeePerStep check

* Sponsored fee for Continuation

* Change extraFeePerStep type to long

* Restore tests

* Improve fill of continuation

* Refactor invoke resolving

* Count scripted addresses

* Prioritize Continuation and Invoke by extra fee per step

* Improve UTX

* Avoid processing Continuation Transaction Diff out of turn

* Refactor for rollback

* Support continuation rollback

* Test full rollback

* Improve rollback test

* Avoid missing of Continuation Transaction when processing

* SC-663-api-add-contiuations-list (#3275)

* SC-663-api-add-contiuations-list

* Update node/src/main/scala/com/wavesplatform/api/http/TransactionsApiRoute.scala

Co-authored-by: Artyom Sayadyan <xrtm000@gmail.com>

* Update node/src/main/scala/com/wavesplatform/api/http/TransactionsApiRoute.scala

Co-authored-by: Artyom Sayadyan <xrtm000@gmail.com>

* SC-663-api-add-contiuations-list
Enrich more APIs.

* SC-663-api-add-contiuations-list
Some test.

* SC-663-api-add-contiuations-list
V3 transaction is supperted.

* SC-663-api-add-contiuations-list
Add call field.

* SC-663-api-add-contiuations-list
Add dApp field.

* SC-663-api-add-contiuations-list
Add feeIncreaseFactor

* SC-663-api-add-contiuations-list
Use raw integer for fixed pointer feeIncreaseFactor in API.

* SC-663-api-add-contiuations-list
Fix merging.

* SC-663-api-add-contiuations-list
Add version and feeAssetId fields.

* SC-663-api-add-contiuations-list
Merge

* Fix continuationTransactionIds typo

* Use V3 Invoke for failed continuation suite

* SC-663-api-add-contiuations-list
Test

* SC-663-api-add-contiuations-list
merging

* SC-663-api-add-contiuations-list
Allow V2

* SC-663-api-add-contiuations-list
Some checks

* SC-663-api-add-contiuations-list
Use V3 transactions

Co-authored-by: Mikhail Potanin <mpotanin@wavesplatform.com>
Co-authored-by: Artyom Sayadyan <xrtm000@gmail.com>

* Allow objects on read of continuation states

* Improve suite

* Add continuation fields to responses for all API methods

* Remove check of Continuation whitelist

* Implement partial rollback

* Optimize imports

* Correct ContinuationSuite

* Await for solidification of block in suite

* Prioritize Continuation as Invoke

* Restore Continuation whitelist checking

* Correct definition of TransactionsOrdering.InBlock

* Add continuation prioritization test

* Improve resolve of Invoke for Continuation

* Use UTX transactions in TransactionsOrdering

* Improve check of Continuation absence in UTX

* Correct ContinuationSuite assertions

* Improve naming

* Correct unit test

* Implement step by step write-off of fee

* Improve fill of continuation nonces field

* More easy-read fill of continuation nonces field

* Implement and test fee distribution

* Improve check of caller balance

* Improve fetch of continuations for API

* Correct assert of balance

* Correct assert of balance

* Correct definition of continuationStatesCache

* Don't add continuations to UTX pool

* Improve continuation suite checks

* Refactor access to estimator instances

* Optimize imports

* Provide backward compatibility for transaction application status

* Move application status cases into object

* Correct tests after refactoring

* Refactor TransactionsOrdering

* Move continuation blocking check to CommonValidation

* Short define of specificCases of blocking by continuation

* Move serialization of ContinuationTransaction to it class

* Move resolveInvoke method to BlockchainExt

* Remove utxTransactions from TransactionOrdering

* Optimize imports

* Use UTX transactions to find Invoke for Continuation

* Add RideV5ActivationSuite

* Get rid of max continuation steps

* Test Invoke extraFeePerStep field

* Correct read transaction status

* Use corrected time as Continuation timestamp

* Resolve invoke using block transactions in Block Differ

* Implement updated application status semantics

* Remove printlns

* Decrease activation height

* Refactor use of replacingTransactions in LevelDBWriter

* Correct write of address transactions

* Improve field names and remove unused

* Correct use of InProgress constructor

* Rename remaining complexity to unused complexity

* Disallow continuation if Invoke version below V3

* Test set script with too many state calls

* Use original height for store of invoke script results

* Correct fee portfolios for miner

* Use DApp returning actions and more accurate check of state changes and balances

* Remove redundant mock from InvokeScriptTransactionDiffTest

* Remove transactionId field from Finished status

* Remove unused RideV5 suite tests

* Rename applied to status

* Add ContinuationTransactionDiffTest

* Optimize imports

* Provide TracedResult destructuring in for-comprehension

* Imrpove ContinuationTransactionDiffTest

* ContinuationDiffTest with account-script

* Correct InvokeScriptTransactionDiffTest

* Correct lang tests

* Improve set of scriptsRun and scriptsComplexity

* Restrict StdLibVersion for Continuation

* Count invoke payments at last step

* Correct resolve of stepInfo

* Improve code

* Use V5 as default version for /utils after activation of ContinuationTransaction

* Correct message for 'can't set continuation before activation' suite

* Allow /utils/compileCode and /utils/estimate for big DApps from V5

* Add doc for checkContinuation param and improve naming

* Support stateCallsComplexities for JS API

* Add 'disallow continuation dApps for version below V5' unit test

* Remove dummy estimator from tests

* Test CommonValidation

* Remove redundant field names

* Improve TransactionsOrderingSpecification namings

* Invalidate continuationStatesCache on InvokeScriptTransaction rollback

* Add rollback test

* Reformat CommonValidationTest

* Restore dummyEstimator for ContinuationSuite

* Optimize imports

* Add UtxPool ordering test

* Count estimated complexity on continuation error

* Add unit test for failed continuation

* Add test for failed by action asset verifier continuation

* Test failed by payment asset verifier continuation

* Add verify of continuation payments

* Additional parens to avoid compile error

* Adapt merged tests

* More predictable use of random in ContinuationTransactionDiffTest

* Provide ordered storage and write of continuation states to avoid inconsistency

* Correct ContinuationTransactionDiffTest

* Correct resolve current continuation step in LevelDB

* Remove timestamp

* Adapt tests

* Clean-up imports

* Ignore Continuation chainId on serialize

* Ignore additional continuation fields if invoke can't be resolved

* Reduce DB read operations

* Remove throw exception in blockchain method

* Remove timestamp field

* Implement parseBytes for ContinuationTransaction

* Support absence of tx timestamp in integration tests

* Bump version to 1.3.0

* Refactor pass of Utx.resolveInvoke to TransactionsOrdering

* Put read continuation states from DB in one batch

* Mark tx as validated in UTX while it is blocked by continuation

* Optimize imports

* Check continuations again after end of UTX loop

* Clean-up UTX tests

* Use UTX PackResult diff to generate continuations

* Use actual PackResult for continuations

* Add test

* Correct fill of block complexity on fail of ContinuationTransactionDiff

* Reduce Continuation JSON body

* Adapt integration tests

* Optimize imports

* Correct ContinuationTransactionDiffTest

* Write invoke ID with continuation state

* Adapt ContinuationSuite

* Store continuation states by address

* Read all continuation states from db

* Correct continuation prioritization test

* Correct activationHeight

* Reduce Continuation fields and restore continuationTransactionIds

* Store all states in db

* Multiple continuations for one dApp test

* Check payments in first step

* Optimize imports

* Freeze of fee

* Correct unit tests

* Disallow unordered continuations

* Refactor EvaluatorV2

* Remove AssetsApiRouteSpec.scala

* Reduce read from db in loadContinuationStates()

* Keep in diff only last state

* Correct check of invoke status on append

* Avoid ContinuationTransactionDiff error on inconsistency of step

* Provide ability of multiple continuation rollbacks for same dApp

* Simplify store of continuation history

* Correct tests

* Correct loadContinuationStates()

* Improve rollback for continuation cache

* Remove invokeScriptTransactionId field from Finished class and convert it to object

* Refactor LevelDB

* Add threshold for continuation states

* Unordered step

* Adapt integration test

* Adapt count of continuation steps

* Rename step to nonce

* Fix get id of address when checking invoke

* Fix compile errors

* Clean-up imports

* Correct imports

* Better naming

* Fill timestamp for Continuation transaction

* Optimize UtxPoolImpl.nonPriorityTransactions

* Use CompositeBlockchain in UtxPool for continuations

* Rename ContinuationStateKey

* Remove ContinuationState destructure

* Use immutable map for Blockchain interface

* Correct API

* Correct check if invoke is in progress

* Improve store and access for continuation state

* Refactor ContinuationTransactionDiffTest

* Remove continuationTransactionIds from Blockchain interface

* Correct usage of ContinuationHistory

* Refactor clear continuation history

* Use protobuf model for continuation history

* Improve updateContinuationCache

* Simplify ContinuationHistory format

* Use diff for continuationTransactionIds API

* Use mutable reference instead of mutable map in Caches

* Rename continuationTransactionIds db key to ContinuationTransactionsHeightsAndNums

* Refactor resolveInvoke(ContinuationTransaction)

* Clean-up imports

* Adapt unit tests

* Map IncompleteResult to empty InvokeScriptResult to avoid errors

* Use diff transactions to find continuation invoke

* More verbose error if can't find InvokeScriptTransaction

* Use collections views in BlockDiffer

* Remove unused config from ContinuationSuite

* Set ContinuationTransaction feature implemented

* Avoid throw exception when resolve invoke

* Avoid generating continuations on UTX clean

* Refactor checkForContinuations flag

* Better naming

* Don't check tx for continuation block on UTX clean

* Adapt unit tests with new InvokeScriptTransaction signature

* Correct signature of ContinuationSuite

* Better naming

* Restore InvokeScriptTransactionDiffTest

* Resolve original logic

* Correct test files definition

Co-authored-by: Mike Potanin <mpotanin@gmail.com>
Co-authored-by: Mikhail Potanin <mpotanin@wavesplatform.com>

Commit:
SC-687-crosscontract-invoke
Disable use Invoke with continuations.

xrtm000 committed:
Avoid adding continuation transaction on UTX clean-up

Commit:
SC-687-crosscontract-invoke
Separate callableV5ReturnType and callableV4ReturnType

Commit:
SC-687-crosscontract-invoke
Fix misstype

Commit:
SC-687-crosscontract-invoke
protobuf dependensies changed

Commit:
Merge remote-tracking branch 'origin/version-1.3.x' into SC-687-crosscontract-invoke

Commit:
SC-687-crosscontract-invoke
Full combine function for InvokeScriptResult.

Commit:
SC-687-crosscontract-invoke
Fix comparation

Commit:
SC-687-crosscontract-invoke
Check all issues for fee calkulating.

Commit:
SC-687-crosscontract-invoke
Fix

xrtm000 committed:
Merge remote-tracking branch 'origin/version-1.3.x' into SC-687-crosscontract-invoke

xrtm000 committed:
Merge branch 'SC-687-crosscontract-invoke' of github.com:wavesplatform/Waves into SC-687-crosscontract-invoke

xrtm000 committed:
Correct fill of protobuf InvokeScriptResult

xrtm000 committed:
Correct fee for crosscontract call with issue test

xrtm000 committed:
Use dependency without leases
